File: Graph Algorithm/python/labyrinth.py
"""
    Author: soju-jsh
    Created: 03/07/24
"""

from sys import stdin, stdout
input = stdin.readline
# print = stdout.write

'''
    Labyrinth
    https://cses.fi/problemset/task/1193
    
    Description:
        >   You are given a map of a labyrinth, 
            and your task is to find a path from start to end.
            
        >   You can walk left, right, up and down.
'''

from collections import deque

def neighbor(pos, mapa, n, m):
    x, y = pos
    
    dx = [0, 0, -1, 1]
    dy = [-1, 1, 0, 0]
    
    nbs = []
    for _ in range(4):
        new_x = x + dx[_]
        new_y = y + dy[_]
        
        if 0 <= new_x < m and 0 <= new_y < n:
            if mapa[new_y][new_x] != '#':
                nbs.append((new_x, new_y))
                
    return nbs

def trace_path(path_to, traces):
    path = []
    x, y = path_to
    cnt = 0
    while traces[y][x] is not None:
        (x, y), _ = traces[y][x]
        path.insert(0, _)
        cnt += 1
    
    print("YES")
    print(cnt)
    print(*path, sep='')

def ulod(pos, mapa, n, m):
    stack = deque()
    stack.append(pos)
    
    vis = [[False for j in range(m)] for i in range(n)]
    prev = [[None for j in range(m)] for i in range(n)]
    movement = {
         0 : { 1 : 'R', -1 : 'L' }, 
         1 : { 0: 'D' }, 
        -1 : { 0: 'U' }
    }
    
    x, y = pos
    vis[y][x] = True
    while stack:
        x, y = stack.popleft()
        for xx, yy in neighbor(pos=(x, y), mapa=mapa, n=n, m=m):
            if not vis[yy][xx]:
                vis[yy][xx] = True
                stack.append((xx, yy))
                prev[yy][xx] = (x, y), movement[yy - y][xx - x]
                
                if mapa[yy][xx] == 'B': 
                    trace_path(path_to=(xx, yy), traces=prev)
                    return
    
    print("NO")
    
        
def main():
    n, m = map(int, input().split())
    mapa = []
    
    # Find Start
    start = 0, 0
    found_start = False
    for i in range(n):
        mapa.append(input().strip())
        if not found_start and 'A' in mapa[i]:
            start = mapa[i].index('A'), i
            
    ulod(pos=start, mapa=mapa, n=n, m=m)

if __name__ == "__main__":
    main()

File: Graph Algorithm/labyrinth.cpp
 /*
    Author: soju-jsh
    Created: 03/07/24
*/

#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using pll = pair<ll, ll>;

/*
    Labyrinth
    https://cses.fi/problemset/task/1193
    
    Result: AC
*/

vector<pll> neighbors(vector<string> &mapa, pll position, ll rows, ll cols){
    vector<int> dx = {0, 0, 1, -1};
    vector<int> dy = {1, -1, 0, 0};
    vector<pll> nbs;

    for (int i = 0; i < 4; i++){
        ll newX = position.first + dx[i];
        ll newY = position.second + dy[i];
        
        if (newX >= 0 && newX < cols && newY >= 0 && newY < rows){
            if (mapa[newY][newX] != '#'){
                nbs.push_back(make_pair(newX, newY));
            }
        }
    }

    return nbs;
}

void trace_path(pll from, pll to, vector<vector<pll>> &trace){
    map<int, map<int, char>> movement = {
        { 0, {{1, 'R'}, {-1, 'L'}}},
        { 1, {{0, 'D'}}},
        {-1, {{0, 'U'}}}
    };

    ll cx = to.first;
    ll cy = to.second;

    string path = "";
    ll degree = 0;

    while (true)
    {
        path += movement[cy - trace[cy][cx].second][cx - trace[cy][cx].first];
        ll ccx = cx;
        cx = trace[cy][cx].first;
        cy = trace[cy][ccx].second;
        ++degree;

        if(cx == from.first && cy == from.second)
            break;
    }

    cout << degree << '\n';
    reverse(path.begin(), path.end());
    cout << path;

    return;
}

void dfs(vector<string> &mapa, pll position, pll target, ll rows, ll cols){
    vector<vector<int>> vis(rows, vector<int>(cols));
    vector<vector<pll>> trace(rows, vector<pll>(cols));

    deque<pll> line;
    line.push_back(position);

    bool rchble = false;
    while (!line.empty()) {
        pll pos = line.front();
        line.pop_front();

        if (!vis[pos.second][pos.first]){
            vis[pos.second][pos.first] = 1;
            
            for (pll nb : neighbors(mapa, pos, rows, cols)){
                if(!vis[nb.second][nb.first]){
                    line.push_back(nb);
                    trace[nb.second][nb.first] = pos;

                    if (nb.first == target.first and nb.second == target.second){
                        rchble = true;
                        line.clear();
                        break;
                    }
                }
            }
        }
    }

    if (!rchble){
        cout << "NO" << endl;
        return;
    } else {
        cout << "YES\n";
        trace_path(position, target, trace);
        return ;
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    ll n, m; cin >> n >> m;
    vector<string> mapa(n);

    pll start, end;
    for (ll i = 0; i < n; i++){
        cin >> mapa[i];
        for (ll j = 0; j < m; j++){
            if(mapa[i][j] == 'A')
                start = make_pair(j, i);
            if(mapa[i][j] == 'B')
                end = make_pair(j, i);
        }
    }

    dfs(mapa, start, end, n, m);
}